marklogic-data-hub/src/main/resources/ml-modules/root/com.marklogic.smart-mastering/matcher-impl/matcher-impl.xqy

```diff
@@ -355,20 +373,84 @@ declare function match-impl:query-map-to-query(
       return (
         if (fn:exists($query-prov)) then
           for $query-hash in document {$query}//schema-element(cts:query) ! xdmp:md5(document{.})
-          where fn:not(map:contains($query-prov,$query-hash)) or ((map:get($query-prov, $query-hash) => map:get("type")) = "reduce")
+          where fn:not(map:contains($query-prov, $query-hash)) or ((map:get($query-prov, $query-hash) => map:get("type")) = "reduce")
           return map:put($query-prov, $query-hash, $sub-query-map)
         else (),
         $query
+      ),
+
+      match-impl:multi-struct-prop-multi-value-queries($query-map, $document, $target-entity)
       )
     (: We want to be certain that didn't lose any queries in the match ruleset, since we don't call the function if no values exist :)
-    where fn:exists($queries) and fn:count($queries) eq fn:count($sub-query-maps)
+    (: FSnow: need to remove the count check since we are returning one query for multiple properties in the structured property case :)
+    where fn:exists($queries) (: and fn:count($queries) eq fn:count($sub-query-maps) :)
     return (
-      map:put($cached-queries, $query-map => map:get("matchRulesetId"), $queries),
+      (: no caching in the structured properties case (yet) :)
+      if (fn:not(map:contains($query-map, "multiStructPropMultiValueMap"))) then
+        map:put($cached-queries, $query-map => map:get("matchRulesetId"), $queries)
+      else
+        ()
+      ,
       $queries
     )
 
 };
 
+declare function match-impl:multi-struct-prop-multi-value-queries($query-map, $document, $target-entity as xs:string?)
+  as cts:query*
+{
+  let $multi-struct-prop-multi-value-map := map:get($query-map, "multiStructPropMultiValueMap")
+  return
+    if (fn:empty($multi-struct-prop-multi-value-map) or fn:empty($target-entity)) then
```

Comment:
When I have an empty if or else, I like to do a where/return to make it a little easier to read - e.g. 

```
let ...
where fn:exists(the map) and fn:exists(target-entity-type)
return 
```